From 93da1a5b85dbca91739cc33edab0bd00e7988194 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Tue, 28 May 2024 09:33:11 +0800
Subject: [PATCH 123/277] mtk: mt76: mt7996: remove chanctx in mt7996_bss_conf

Different vif under the same phy cannot use different chanctx.
So it is better to maintain the chanctx in mt7996_phy.
Instead of mconf->chanctx, get the chanctx by mconf->phy->chanctx.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt7996/main.c   | 8 +-------
 mt7996/mt7996.h | 2 --
 2 files changed, 1 insertion(+), 9 deletions(-)

diff --git a/mt7996/main.c b/mt7996/main.c
index 5a6165b5..c8dc0687 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2539,8 +2539,6 @@ mt7996_assign_vif_chanctx(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		return ret;
 	}
 
-	mconf = mconf_dereference_protected(mvif, link_id);
-	mconf->chanctx = ctx;
 	ctx->nbss_assigned++;
 	mvif->band_to_link[phy->mt76->band_idx] = link_id;
 
@@ -2559,8 +2557,6 @@ mt7996_unassign_vif_chanctx(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 {
 	struct mt7996_chanctx *ctx = mt7996_chanctx_get(conf);
 	struct mt7996_phy *phy = ctx->phy;
-	struct mt7996_vif *mvif = (struct mt7996_vif *)vif->drv_priv;
-	struct mt7996_bss_conf *mconf;
 
 	wiphy_info(hw->wiphy, "Remove VIF (addr: %pM, type: %d, link_id: %d) from channel context: %d MHz\n",
 		   vif->addr, vif->type, link_conf->link_id,
@@ -2572,8 +2568,6 @@ mt7996_unassign_vif_chanctx(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 	if (test_bit(MT76_SCANNING, &phy->mt76->state))
 		mt7996_scan_complete(phy, true);
 
-	mconf = mconf_dereference_protected(mvif, link_conf->link_id);
-	mconf->chanctx = NULL;
 	ctx->nbss_assigned--;
 
 	mutex_unlock(&phy->dev->mt76.mutex);
@@ -2747,7 +2741,7 @@ mt7996_event_callback(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 
 				/* trigger calibration for DFS link */
 				if (!cfg80211_reg_can_beacon(hw->wiphy,
-							     &mconf->chanctx->chandef,
+							     &phy->chanctx->chandef,
 							     NL80211_IFTYPE_AP))
 					mt7996_mcu_set_chan_info(phy, UNI_CHANNEL_SWITCH,
 								 true);
diff --git a/mt7996/mt7996.h b/mt7996/mt7996.h
index 07c72d2b..9f58a793 100644
--- a/mt7996/mt7996.h
+++ b/mt7996/mt7996.h
@@ -348,8 +348,6 @@ struct mt7996_bss_conf {
 	struct ieee80211_tx_queue_params queue_params[IEEE80211_NUM_ACS];
 	struct cfg80211_bitrate_mask bitrate_mask;
 
-	struct mt7996_chanctx *chanctx;
-
 	u8 link_id;
 	u8 own_mld_id;
 	u8 bpcc;
-- 
2.45.2

